edge-modules/api-proxy-module/src/monitors/certs_monitor.rs

```diff
@@ -0,0 +1,327 @@
+            }
+        }
+    });
+
+    Ok((monitor_loop, shutdown_handle))
+}
+
+struct CertificateMonitor {
+    module_id: String,
+    generation_id: String,
+    hostname: String,
+    bundle_of_trust_hash: String,
+    work_load_api_client: edgelet_client::WorkloadClient,
+    cert_expiration_date: DateTime<Utc>,
+    validity_days: Duration,
+}
+
+impl CertificateMonitor {
+    pub fn new(
+        module_id: String,
+        generation_id: String,
+        hostname: String,
+        workload_url: String,
+        validity_days: Duration,
+    ) -> Result<Self, Error> {
+        //Create expiry date in the past so cert has to be rotated now.
+        let cert_expiration_date = DateTime::parse_from_rfc3339(EXPIRY_TIME_START_DATE)
+            .context("Error reading Start date")?
+            .with_timezone(&Utc);
+
+        let work_load_api_client = match edgelet_client::workload(&workload_url) {
+            Ok(work_load_api_client) => (work_load_api_client),
+            Err(err) => {
+                return Err(anyhow::anyhow!(format!(
+                    "Could not get workload client {:?}",
+                    err
+                )))
+            }
+        };
+
+        Ok(CertificateMonitor {
+            module_id,
+            generation_id,
+            hostname,
+            bundle_of_trust_hash: String::from(""),
+            work_load_api_client,
+            cert_expiration_date,
+            validity_days,
+        })
+    }
+
+    async fn need_to_rotate_server_cert(
+        &mut self,
+        current_date: DateTime<Utc>,
+    ) -> Result<(bool, String, String), anyhow::Error> {
+        let mut need_to_rotate = false;
+        let server_crt;
+        let private_key;
+
+        if current_date >= self.cert_expiration_date {
+            let new_expiration_date = Utc::now()
+                .checked_add_signed(self.validity_days)
+                .context("Could not compute new expiration date for certificate")?;
+            let resp = self
+                .work_load_api_client
+                .create_server_cert(
+                    &self.module_id,
+                    &self.generation_id,
+                    &self.hostname,
+                    new_expiration_date,
+                )
+                .await?;
+
+            server_crt = resp.certificate().to_string();
+            let private_key_raw = resp.private_key();
+            private_key = match private_key_raw.bytes() {
+                Some(val) => val.to_string(),
+                None => return Err(anyhow::anyhow!("Private key field is empty")),
+            };
+            need_to_rotate = true;
+
+            let datetime = DateTime::parse_from_rfc3339(&resp.expiration())
+                .context("Error parsing certificate expiration date")?;
+            // convert the string into DateTime<Utc> or other timezone
+            self.cert_expiration_date = datetime.with_timezone(&Utc);
+        } else {
+            server_crt = "".to_string();
```

Comment:
`None` instead??

Reply:
the pattern can be like this 
```rust
if current_date >= self.cert_expiration_date {
  ...
  (true, Some(cert, pkey))
} else {
  (false, None)
}
```
after this change above it feels like `bool` is redundant. rust type system is pretty powerful. we should leverage it!

Reply:
You are totally right. I refactored that part.